fix(forms): ValueAccessor.writeValue(...) sometimes called with null

N2D4 opened the pull request:
Fixes #14988
Fixes [@ckelley7's marriage](https://github.com/angular/angular/issues/14988#issuecomment-396342060)
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
(technically no, but no doc changes were needed)

## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
`ValueAccessor.writeValue(...)` is called twice on the very first `ngModel` change if the form has not yet been set up; once by `setUpControl` and once when the value is actually updated. However, because `setUpControl` is called before the value is set, the first call passes `null`.

The bug is most common for standalone form controls which are never set up before the first `ngModel` change but may also happen if any other form control is not added to its parent first by calling `NgForm.addControl(...)`.

Issue Number: #14988


## What is the new behavior?
Optionally skip the `writeValue(...)` call in `setUpControl` if the value is a special sentinel value indicating it has not yet been initialized.

## Does this PR introduce a breaking change?

- [x] Yes
(see [below](https://github.com/angular/angular/pull/38140#issuecomment-661516210))
- [ ] No


## Other information

N2D4 committed:
fix(forms): ValueAccessor.writeValue(...) sometimes called with null

`ValueAccessor.writeValue(...)` is called twice on the very first `ngModel` change if the form has
not yet been set up; once by `setUpControl` and once when the value is actually updated. However,
because `setUpControl` is called before the value is set, the first call passes `null`.
The bug is most common for standalone form controls which are never set up before the first
`ngModel` change but may also happen if any other form control is not added to its
parent first by calling `NgForm.addControl(...)`.
Fix that by optionally skipping the `writeValue(...)` call in `setUpControl` if an optional flag
is given.

Fixes #14988

AndrewKushnir commented:
Hi @N2D4, thanks for creating this PR 👍 

I noticed that it's in Draft mode (i.e. it's probably not fully ready for review), but I wanted to share an idea. If I understand correctly the first "writeValue" with the `undefined` value comes as a result of `FormControl` instance creation in `ng_model.ts` [here](https://github.com/angular/angular/blob/master/packages/forms/src/directives/ng_model.ts#L137) (with no value). I was thinking if we can instantiate `FormControl` there with some "initial value" and check for it before calling the `writeValue` method in the `setUpControl` function, i.e. something like this:

```
// in packages/forms/src/directives/ng_model.ts:

export const INITIAL_CONTROL_VALUE = {};

...

public readonly control: FormControl = new FormControl(INITIAL_CONTROL_VALUE);


// in packages/forms/src/directives/shared.ts:

if (control.value !== INITIAL_CONTROL_VALUE) {
  dir.valueAccessor.writeValue(control.value);
}
```
Would that work? (If that works as needed, it should avoid passing configuration params though a few function calls)

Thank you.

N2D4 commented:
> Hi @N2D4, thanks for creating this PR 👍
> 
> [...]
> 
> Would that work? (If that works as needed, it should avoid passing configuration params though a few function calls)
> 
> Thank you.

Sure, implemented. I also added an integration test (/modified an existing one), so this is now ready for review 👍

alfaproject commented on packages/forms/src/directives/shared.ts:
```diff
@@ -27,6 +27,15 @@ import {SelectControlValueAccessor} from './select_control_value_accessor';
+ * @internal
+ */
+export const INITIAL_CONTROL_VALUE = {
+  _sentinel: ['shared/INITIAL_CONTROL_VALUE']
```
There's no need to increase bundle size with this extra attribute. The object reference is enough and you can just add a comment here instead if you want

N2D4 replied:
The logic is that if anyone messes around with the code elsewhere and doesn't see where the value is coming from they will have an easier time finding it, but I'll change it regardless

AndrewKushnir commented:
Hi @N2D4,

I performed initial review and added the "breaking changes" flag to this PR, since the new behavior is different from the old one (i.e. there us no additional `writeValue` call) and developers may rely on that, thus we need to highlight this change in the Breaking Changes section of Release Notes.

I've also ran a set of tests in Google's codebase ([internal-only link](https://test.corp.google.com/ui#id=OCL:322189886:BASE:322189965:1595270425941:46fe1c85)) and it looks like the problem is a bit more tricky. I found some additional cases were this change would be breaking. Here is a couple snippets that illustrate the problem:

Component's template:
```
<input
  myValidator
  [(ngModel)]="someField"
/>
```

Validator (simplified version):
```
@Directive({
  selector: '[myValidator]',
  providers: [{provide: NG_VALIDATORS, useExisting: MyValidator, multi: true}]
})
export class MyValidator implements Validator {
  validate(control: AbstractControl) {
    console.log(control.value);
    return null;
  }
}
```

The problem is that the `control.value` value in the `validate` function would be `INITIAL_CONTROL_VALUE`, i.e. just an empty object, thus causing some code to fail (the code that may not expect to see the object there at all). Also from API perspective having the `{}` here is quite unexpected. Also the question is whether a validator (sync/async) should actually be invoked in that case as well (I believe it should not)?

Given the fact that the `INITIAL_CONTROL_VALUE` may leak in some other places (possibly in `ngOnChanges` hook if you use `ngModel` as an input on a custom element, like `<my-comp [(ngModel)]="someField">`), I'm wondering if we should go back to the flags-based approach.

In any case, I think it'd be great to add the above mentioned example as a test case (it should be useful to count the number of times sync/async validator was called) and evaluate whether `INITIAL_CONTROL_VALUE`-based approach may work or other solution should be explored.

Please let me know your opinion.

Thank you.

N2D4 commented:
Hi @AndrewKushnir,

Sorry for the delay, since this is scheduled for v11 I ended up re-prioritizing things but I'm back to work on this.

As an open-source intern I don't have access to the internal link, but I could manage to reproduce the error you mentioned. I agree that the validator shouldn't be called there, and added a unit test to make sure it's not. Yes, I believe in the `ngOnChanges` example you mentioned the value would leak as well, so I reverted back to the flag approach. Thanks for the feedback.

Now that we also skip validators when the form control is created, I took a closer look and decided to rename the flag to `skipValueInitialization` as that fits it better semantically. It should be passed to `addControl` (and the other methods) if and only if the value will be set/updated/initialized immediately after the call (as compared to already having been initialized). I've been thinking that we could also pass an initialization value instead - if given, `addControl` itself would initialize the control with that value first. It would complicate some of the logic in `setupControl` and `addControl` by a bit, but removes that weird invariant of "if you pass this flag, you MUST update the value immediately". Let me know if you find one better than the other.

There was one test that relied on the old behavior of calling validators synchronously during `ngOnChanges` when the component is first created. The call with argument `null` was already asynchronous for non-standalone, but synchronous for standalone components. Since the `null` call no longer happens, no such synchronous validator call will happen anymore. I don't think this should cause a lot of breakage in consumer unit tests, but I still thought it's worth pointing out.

JounQin commented:
Then, any news about merging?

JounQin commented on packages/forms/src/directives/shared.ts:
```diff
@@ -27,20 +27,28 @@ import {SelectControlValueAccessor} from './select_control_value_accessor';
+ *
+ * @internal
+ */
+export const INITIAL_CONTROL_VALUE = {};
```
It is not used anywhere anymore.

alxhub commented:
This fix requires some additional design work / research as the current approach is too breaking to land as-is.
